create fallback turn server option

CicadaCinema opened the pull request:
Fixes #2329 
Fixes #3063
You need to enable the fallback option in `Settings > Voice & Video` to use this feature:
![Screenshot_20211106-215730_Element dbg.png](https://user-images.githubusercontent.com/52425971/140624995-a3804042-6519-414a-9dc8-efe6c3e0f639.png)

### Pull Request Checklist

<!--
 Please read [CONTRIBUTING.md](https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md) before submitting your pull request
 Depending on the Pull Request content, it can be acceptable if some of the following checkboxes stay unchecked.
 -->

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

bmarty commented on vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCall.kt:
```diff
@@ -269,17 +271,28 @@ class WebRtcCall(
         val iceServers = mutableListOf<PeerConnection.IceServer>().apply {
-            turnServerResponse?.let { server ->
-                server.uris?.forEach { uri ->
+            if (vectorPreferences.useFallbackTurnServer()) {
```
I'm not a specialist on this part of the code, but maybe use the fallback if and only if `turnServerResponse?.uris is null or empty, and the user has enable the fallback from the setting?

bmarty commented on vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCall.kt:
```diff
@@ -269,17 +271,28 @@ class WebRtcCall(
-            turnServerResponse?.let { server ->
-                server.uris?.forEach { uri ->
+            if (vectorPreferences.useFallbackTurnServer()) {
+                listOf("turn:turn.matrix.org:3478?transport=udp", "turn:turn.matrix.org:3478?transport=tcp", "turns:turn.matrix.org:443?transport=tcp").forEach {
```
Maybe extract this list to the file https://github.com/vector-im/element-android/blob/main/vector/src/main/res/values/config.xml#L2 ?

bmarty reviewed (commented):
Thanks for the PR.
Please see my first remarks.
Also can you add a file for the changelog as per https://github.com/vector-im/element-android/blob/main/CONTRIBUTING.md#changelog ?

Thanks

CicadaCinema commented:
Thank you for reviewing this! I actually don't know if this change fixes #2329 at all, I've asked people there to test out my builds. I'll finish up work on this PR if I get positive feedback on the changes.

bmarty commented:
Ok, thanks. I have enabled the CI for this PR so that APKs will be built and can be downloaded from here to be tested.

CicadaCinema commented:
Thank you!

DoM1niC commented:
Nice PR @CicadaCinema but I prefer first to use the TURN from Homeserver instead of general Matrix to use.... I my mind you can read this from API or you switch the Server URLs to gradle vars to find and setup faster the own TURN Server like me....

But anyway good work 👍🏻

CicadaCinema commented:
> Nice PR @CicadaCinema but I prefer to use the TURN first from Homeserver instead of general Matrix to use.... I think, you can read this from API or you switch the Server URLs to gradle vars to find and setup faster the own TURN Server like me....
> 
> But anyway good work 👍🏻 

If this setting is enabled, I plan to _only_ use turn.matrix.org when the user's homeserver does not have one that they can use (similarly to the way the setting in Element Web works).

CicadaCinema commented:
Signed-off-by: Arseny Uskov <arsenyuskov@gmail.com>

ouchadam commented:
awesome change! with the fallback only being used when no turn server uris are returned by the server **and** the preference is enabled this looks like it ticks everyone's boxes

bmarty commented on vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCall.kt:
```diff
@@ -280,13 +285,15 @@ class WebRtcCall(
-                server.uris?.forEach { uri ->
+                val useFallback = server.uris.isNullOrEmpty() && vectorPreferences.useFallbackTurnServer()
+                val serverList = if (useFallback) stringArrayProvider.getStringArray(R.array.fallback_ice_servers).toList() else server.uris
+                serverList?.forEach { uri ->
```
Maybe a stupid question, but do we want to use the fallback if `turnServerResponse` is null? Here this will not be the case.

BillCarsonFr replied:
Yes we should only use the fallback if the home server does not define any ice servers (consitent with web).

bmarty replied:
Ok, so the code `turnServerResponse?.let { server ->` has to be updated IIUC.

CicadaCinema replied:
there, I've handled that now

BillCarsonFr commented:
Notice that using a fallback while revealing your IP to the relay will also hide it to your contact.
In other app this option can be presented as always relay Calls (to avoid revealing your IP adress to your contact).
So the current text copy is maybe a bit too negative

BillCarsonFr commented:
Also the old android app was prompting the user to enable this option when you have a call that failed to connect and that your home server had no ice server config. Maybe a nice to have feature

BillCarsonFr commented on vector/src/main/java/im/vector/app/features/call/webrtc/WebRtcCall.kt:
```diff
@@ -280,13 +285,15 @@ class WebRtcCall(
                                     .builder(uri)
-                                    .setUsername(server.username)
-                                    .setPassword(server.password)
+                                    .setUsername(if (useFallback) "xxxxx" else server.username)
```
The fallback is stun, should do (from old sdk)
````
 if (!defaultIceServerUri.startsWith("stun:")) {
                    defaultIceServerUri = "stun:" + defaultIceServerUri;
                }
                defaultIceServer = PeerConnection.IceServer.builder(defaultIceServerUri).createIceServer();

BillCarsonFr commented on vector/src/main/res/values/config.xml:
```diff
@@ -7,6 +7,12 @@
     <string name="bug_report_url" translatable="false">https://riot.im/bugreports/submit</string>
+    <string name="fallback_stun_server_url" translatable="false">turn.matrix.org</string>
+    <string-array name="fallback_ice_servers" translatable="false">
+        <item>turn:turn.matrix.org:3478?transport=udp</item>
```
I don't think this will work, turn.matrix.org requires auth.
Just use stun: would work

bmarty replied:
`fallback_ice_servers` can just be removed now I think

CicadaCinema replied:
yep, I've removed it

BillCarsonFr commented on vector/src/main/res/values/strings.xml:
```diff
@@ -724,7 +724,7 @@
     <string name="template_settings_call_ringtone_use_app_ringtone">Use default ${app_name} ringtone for incoming calls</string>
     <string name="settings_call_ringtone_use_default_stun">Allow fallback call assist server</string>
-    <string name="settings_call_ringtone_use_default_stun_sum">Will use "%s" as assist when your homeserver does not offer one (your IP address will be shared during a call)</string>
+    <string name="settings_call_ringtone_use_default_stun_sum">Will use %s as assist when your homeserver does not offer one (your IP address will be shared during a call)</string>
```
I think this copy is missleading. IP will be shared with who? the peer yes, the stun yes.
Not using it will still share your IP with the peer.
So maybe your IP will be seen by the stun server? (not sure how valuable/actionable this information is for the user, given that not doing it will make calls fails depending on your network)

BillCarsonFr reviewed (changes requested):
Hello,
The default server will only do stun, please see my comments.

CicadaCinema commented:
@BillCarsonFr 

I adjusted the behaviour of the fallback to use the stun server instead, and I adjusted the option description.

> Also the old android app was prompting the user to enable this option when you have a call that failed to connect and that your home server had no ice server config. Maybe a nice to have feature

I'm not sure how to even begin implementing this unfortunately - I only took a look at this issue because I noticed I couldn't make calls myself.

CicadaCinema commented:
# PR Moved to https://github.com/vector-im/element-android/pull/5781
